pkg/operator/controller/externaldns/deployment.go

```diff
@@ -105,8 +114,18 @@ func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, nsName ty
 	return true, depl, nil
 }
 
+// extractCredentialsSecret retieves the contents of the credentials' secret
+func (r *reconciler) extractCredentialsSecret(ctx context.Context, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Secret, error) {
+	secretName := externalDNSCredentialsSecretName(externalDNS)
+	secret := &corev1.Secret{}
+	if err := r.client.Get(ctx, secretName, secret); err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
```

Comment:
It appears that the only provider-specific logic for credentials is in determining the name of the secret.  Is that correct?

If so, the fact that this logic is almost completely provider-agnostic begs the question of whether the API field needs to be provider-specific (e.g., `Credentials` in `ExternalDNSAWSProviderOptions`) or whether it could be simplified by replacing `Credentials` in `ExternalDNSAWSProviderOptions`, `ExternalDNSGCPProviderOptions`, and `ExternalDNSInfobloxProviderOptions` and `ConfigFile` in `ExternalDNSAzureProviderOptions` and `ExternalDNSBlueCatProviderOptions` with a single `Credentials` field under `ExternalDNSProvider`.  

Is that something worth considering?  Is there some provider where the credentials config is wildly different (in which case presumably this logic will need to be significantly changed)?  (Maybe we already discussed this?)

I suppose one counterargument is that even if each provider needs a credentials/config secret, the keys vary from one provider to the next, so maybe using the same API field for all providers would be misleading.  :thinking:

Reply:
> I suppose one counterargument is that even if each provider needs a credentials/config secret, the keys vary from one provider to the next, so maybe using the same API field for all providers would be misleading.

I wonder if this is the reality. The times I've used cloudflare there's an account key and an API token.

Reply:
1. Infoblox provider has some fields additional to the credentials secret: https://github.com/openshift/external-dns-operator/blob/main/api/v1alpha1/externaldns_types.go#L342-L356
2. You are right @Miciah, the keys in the secret are different for different providers.
3. Also, some providers expect full config file (Azure, GCP, BlueCat) and some others just credentials (AWS, Infoblox). Which means that the potential consolidated name has to be well chosen to work well with both.